### DISCUSSION TITLE
What is zod's approach to conditionally requiring fields?

### DISCUSSION THREAD

**Discussion by ljunio** [colinhacks/zod#938] [Q&A]
I have a situation where I need to conditionally require another field based on the validity of another field. AFAIK, we can only enforce this by using `refine` or `superRefine`. The enforcement however only comes to fruition through the use of `addIssue(...)`. Is there a way to actually set a field as a `required` or `optional` or is that basically what `refine`/`superRefine` does under the hood?

**Comment by scotttrinh** [comment]
> I have a situation where I need to conditionally require another field based on the validity of another field.

You have a few choices, but it depends on what those conditions are. In all things, it's best to be as truthful to the schema as you can, which sometimes requires multiple schemas or some duplication.

Other than `refine`, you might also be able to represent your schema as a `union` where each member has a shape that is valid based on some rules, like:

```ts
const fooSchema = z.object({
  bar: z.string().min(1),
  baz: z.string().min(1),
});

const bazSchema = z.object({
  bar: z.literal(""),
  baz: z.literal(""),
});

const schema = z.union([fooSchema, bazSchema]);

schema.safeParse({ bar: "not empty", baz: "" }); // failure
schema.safeParse({ bar: "", baz: "" }); // success
schema.safeParse({ bar: "not empty", baz: "not empty also" }); // success
```

It might help to provide some simple examples of what problem you're trying to model.

**Comment by JacobWeisenburger** [comment]
There might be other ways to do this, but I feel like [`refine`](https://github.com/colinhacks/zod#refine) is the best way to be the most clear.

Here's a simple example of how to use refine for this.
```ts
const myString = z.object( {
    foo: z.string(),

    // must be optional or parse will fail before refinement is applied
    bar: z.string().optional(),
} ).refine( input => {

    // allows bar to be optional only when foo is 'foo'
    if ( input.foo !== 'foo' && input.bar === undefined ) return false

    return true
} )

// bar is optional here
console.log( myString.safeParse( { foo: 'foo' } ).success ) // true
console.log( myString.safeParse( { foo: 'foo', bar: 'bar' } ).success ) // true
console.log( myString.safeParse( { foo: 'foo', bar: undefined } ).success ) // true

// bar is required here
console.log( myString.safeParse( { foo: 'not foo', bar: 'bar' } ).success ) // true
console.log( myString.safeParse( { foo: 'not foo' } ).success ) // false
console.log( myString.safeParse( { foo: 'not foo', bar: undefined } ).success ) // false
```

**Reply by Aaronius** [reply]
I'm trying to do something similar to @ljunio and followed your example. One problem I have is that I'm trying to merge `myString`, in your example, into another object. Basically:

```
const otherObject = z.object({}).merge(myString);
```

But I can't because ZodEffects isn't assignable to AnyZodObject.

**Reply by Aaronius** [reply]
Similarly, I can't merge a ZodUnion into a ZodObject or a ZodObject into a ZodUnion. Also, this gets really tricky when I try to use ZodEffects or ZodUnions for objects that are part of a ZodDiscriminatedUnion. 😓

**Reply by tkowalski-exxeta** [reply]
This seems to be a bug IMHO. Does any ticket exist for this?

**Reply by gabeodame** [reply]
You can apply superRefine after merging example


```
const mergedSchema = schemaA.merge(schemaB)
export const FormSchema = mergedSchema.superRefine((data, ctx) => {
  if (condition) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      message: "custom message",
      path: ["property"],
    });
  }
});
```

**Reply by bizmich** [reply]
how to make optional if its initially required ?

**Reply by adambak033** [reply]
Thank you really much my friend 🚀

**Comment by toxsick** [comment]
There is also #1394 now

**Comment by ramil-thrive** [comment]
I came up with something like this:

```tsx
const registrationFormSchema = z.object({
    firstName: firstNameValidator,
    lastName: lastNameValidator,
    email: z.string().optional(), // Keeping both email and phone optional on base schema and validating values inside `superRefine`
    phone: z.string().optional(),
    postalCode: canadianPostalCodeValidator,
    specificGroupServices: z.string().optional(),
    isOntarioResident: booleanRadioGroupValidator,
    isAllowedToWorkInCanada: booleanRadioGroupValidator,
    isStudent: booleanRadioGroupValidator,
    hasAgreedToTermsAndPolicies: z.boolean(),
  })
  // Perform conditional validation to ensure either a valid email or phone number is provided.
  .superRefine(({ email, phone }, refinementContext) => {
    if (email === '' && phone !== '' && !getIsValidMobilePhone(phone ?? '')) {
      return refinementContext.addIssue({
        code: z.ZodIssueCode.custom,
        message: i18n.t('forms.validations.invalid.phone'),
        path: ['phone'],
      });
    }

    if (email !== '' && phone === '' && !getIsValidEmail(email ?? '')) {
      return refinementContext.addIssue({
        code: z.ZodIssueCode.custom,
        message: i18n.t('forms.validations.invalid.email'),
        path: ['email'],
      });
    }
  });

```

Example UI:


<img width="420" alt="image" src="https://github.com/colinhacks/zod/assets/132285107/c63ccc73-2a25-4a4a-96bb-b800772a653a">

**Reply by asifsaho** [reply]
perfect, this helped me a lot. I was blocked for days! thanks for sharing :)

**Reply by ashwary23** [reply]
I have another question @ramil-thrive .
lets say I have another schema
`
optionalFieldsSchema: z.object({
    field1: z.string(),
    field2: z.string()
})`
and I want it to be added to **registrationFormSchema** only if isStudent === true
can you let me know the solution, I am blocked since long.

**Reply by micahjon** [reply]
Hey @ashwary23 , I've run into a similar situation before where I want to entirely remove parts of my schema based on a form's conditional logic. The first step is making all conditional fields `.optional()` e.g.
```ts
 optionalFieldsSchema: z.object({ field1: z.string(), field2: z.string() }).optional(),
```

The next part is tricky: you need to detect when this field is hidden and remove it from your form's data *before* validation, so that any validation errors in the hidden field's value doesn't prevent submission (assuming you're using `shouldUnregister = false`). I ran into this again and again, and eventually wrote my own tiny helper library from conditional logic in React Hook Form and Zod:
https://github.com/micahjon/rhf-conditional-logic

Essentially, I just define conditions outside of my Zod schema (for optional fields)...
```ts
const conditions = {
    optionalFieldsSchema: getValues => getValues("isStudent") === true
}
```

And then prune them pre-validation & post-submission. Not an ideal solution but at least this way it's type-safe and declarative, and scales to many conditions, while my schema remains clean & simple.

**Reply by dlibis** [reply]
small correction to your code, return `refinementContext` at the end

**Reply by aman-jha-gsc** [reply]
Perfect this is working for me, thanks for the reply.

**Reply by thinhbg2812** [reply]
It's not working for my issue, i tried condition ``` type !== '0'``` it's still appear the same, no error when i submit form
although value of error is '0'

I need the "JobOpenning" "JobTitle" to be required when type === '0' and optional(because it's will be hidden) when type !='0' :(


![CleanShot 2024-10-29 at 23 23 00@2x](https://github.com/user-attachments/assets/75d98d63-bd62-46cd-b56f-25670eea46d9)


![CleanShot 2024-10-29 at 23 24 24](https://github.com/user-attachments/assets/7b58f636-101e-4ae7-8542-82723105cb21)

**Reply by BrendanC23** [reply]
Can you paste your full code? This kind of approach should work fine. A few notes:

* You can just call `ctx.addIssue()`. You don't need to return it.
* You need to check `type === "0" && !jobTitle`. Right now, you are adding an error every time, not only when a value is missing.
* You can have a separate `addIssue()` call for both `jobTitle` and `jopOpening`

Something like this should work:

```ts
const Schema = z.object({
    type: z.enum(["0", "1"]),
    jobOpening: z.string().optional(),
    jobTitle: z.string().optional(),
});

const Refined = Schema.superRefine((data, ctx) => {
    if (data.type === "0" && !data.jobTitle) {
        ctx.addIssue({
            code: "custom",
            message: "jobTitle is required when type is 0",
            path: ["jobTitle"],
        });

        // Not returning z.NEVER, so execution will continue to the second check
    }
    
    if (data.type === "0" && !data.jobOpening) {
        ctx.addIssue({
            code: "custom",
            message: "jobOpening is required when type is 0",
            path: ["jobOpening"],
        });
    }

    // No need to return ctx
});
```

Another option would be to use a discriminated union, which would give you better type safety and narrowing. This would let you use TypeScript to narrow and only access `jobTitle` and `jobOpening` after you've checked that `type === 0`.

```ts
const Type0Schema = z.object({
    type: z.literal("0"),
    jobTitle: z.string(),
    jobOpening: z.string(),
});

const Type1Schema = z.object({
    type: z.literal("1"),
})

const Schema = z.object({
    value: z.discriminatedUnion("type", [Type0Schema, Type1Schema]),
});
```

With this method, you would need to switch between each schema within an `onChange` event (the below code is based on React-Hook-Form).

```tsx
<Select
    onChange={(option) => {
        if (option === "0") {
            form.setValue("value", {
                type: "0",
                jobTitle: "",
                jobOpening: "",
            });
        } else {
            form.setValue("value", {
                type: "1",
            });
        }
    }}
>
    <option>0 </option>
    <option>1</option>
</Select>
```

**Reply by thinhbg2812** [reply]
Here you are,
i created a new simple project with condition similar to my problem here
i move the zod schema to a separate file, and super refine init 

https://github.com/thinhbg2812-2/react-hook-form-with-condition-validation

**Reply by thinhbg2812** [reply]
After adding the setValue "type"
i manage to get error and block submit
but will only appear after all other field are fulfilled, like "exp" in video 
somehow it will not show at the same time when others conditions are not fulfilled 

I'm trying the discriminatedUnion, thank you for suggestion




https://github.com/user-attachments/assets/e660b69f-187b-4c72-b8e6-16d7c3c71b6f

**Reply by BrendanC23** [reply]
Zod's `refine`/`superRefine` will only run after the schema is fully verified (otherwise, the `data` parameter of `refine` wouldn't match the expected schema and would need to be typed as `unknown`). If you want the refinement to occur even if `experience` hasn't been validated yet, you need to move it to a nested object. See [this thread](https://github.com/colinhacks/zod/issues/479) for more details.

```ts
const commonKeysFormSchema = z.object({
    job: z
        .object({
            type: z.enum([POST_TYPE.SPECIALIZED.toString(), POST_TYPE.GENERAL.toString()]),
            jobOpening: z.string().optional(),
            jobTitle: z.string().optional(),
        })
        .superRefine(), // Call superRefine() on a nested object instead of the entire schema
    experience: z.string({
        message: ZOD_MESSAGES.required,
    }),
});
```

There's an alternative using `z.preprocess()` in [this comment](https://github.com/colinhacks/zod/issues/479#issuecomment-2429834215) but it's more complicated than just adding another level of nesting.

**Reply by thinhbg2812** [reply]
You are the goat sir <3

**Reply by mstosio** [reply]
@ramil-thrive 

> ```tsx
> getIsValidEmail
> ```

This is a fine solution, but you need to add your own utility for validating emails. You can add a custom check here using parse:

```
.superRefine((data, ctx) => {
        const emailCheck = z.string().email();
        const result = emailCheck.safeParse(data.yourEmailField);

        const isEmailRequired = data.yourOtherFieldInSechma === true

        if (isEmailRequired && !result.success) {
            ctx.addIssue({
                path: ['emailToSendApplicationTo'],
                code: z.ZodIssueCode.custom,
                message: 'Email is required when enrollment approval is needed',
            });
        }
    });
```

Is there a better option for adding conditional types than this?

**Comment by dominikjasek** [comment]
I ran into similar issue:

I had these requirements for validation library:

- per-branch type narrowing that's inferred automatically
- the "which field shows when" logic in one place (not smeared across JSX)
- and conditions beyond literals (predicates / cross-field functions).

zod or any other validation libraries were not suitable for this, so I created a `liveschema` package for exactly this use case. It sits right in the middle between form libraries (such as react-hook-form or vee-validate) and validation libraries (such as zod or valibot)

Ended up building [@liveschema/core](https://www.npmjs.com/package/@liveschema/core). [React](https://www.npmjs.com/package/@liveschema/react) and [Vue](https://www.npmjs.com/package/@liveschema/vue) versions have dedicated packages.

Here's code example of it:

```ts

import { defineSchema } from '@liveschema/core'

const schema = defineSchema()
  .field('name', z.string().min(1))
  .field('animal', z.enum(['dog', 'cat']))
  .when({ animal: 'dog' }, (b) => b.field('dogSize', z.enum(['small', 'large'])))
  .when((v) => v.name.length > 20, (b) => b.field('nickname', z.string())) // predicates too
  
```  

Runnable example (React + TanStack Form): https://stackblitz.com/edit/liveschema-react-tanstack?file=src%2Fschema.ts
